persist/storer.go

```diff
@@ -3,168 +3,222 @@ package persist
+// This implementation adds a restriction to the key name: it must be a valid file name (without extension).
+func (j inMemoryStore) Set(key string, value interface{}) int64 {
+	ts := now().Unix()
+	j.cachedData[key] = jsonEntry{
+		Timestamp: ts,
+		Value:     value,
 	}
+	return ts
+}
 
-	data, err := json.Marshal(c.inMemoryStore)
-	if err != nil {
-		return err
+// Get gets the value associated to a given key and stores it in the value referenced by the pointer passed as
+// second argument
+// This implementation adds a restriction to the key name: it must be a valid file name (without extension).
+func (j inMemoryStore) Get(key string, valuePtr interface{}) (int64, error) {
+	rv := reflect.ValueOf(valuePtr)
+	if rv.Kind() != reflect.Ptr || rv.IsNil() {
+		return 0, errors.New("destination argument must be a pointer")
 	}
 
-	return ioutil.WriteFile(c.path, data, filePerm)
-}
+	entry, ok := j.cachedData[key]
 
-// Save won't persist on disk.
-func (c *inMemoryStore) Save() error {
-	return nil
-}
+	// If the entry is not cached, it may be stored as a JSON (as loaded from disk), and we unmarshall it
+	if !ok {
+		bytes, ok := j.Data[key]
+		if !ok {
+			return 0, ErrNotFound
+		}
 
-// Get returns the value for a given key. Last Set call timestamp is returned.
-// If key is not found ErrNotFound error is returned.
-func (c *inMemoryStore) Get(key string) (value interface{}, timestamp int64, err error) {
-	var ok bool
-	if value, ok = c.Data[key]; ok {
-		if timestamp, ok = c.Timestamps[key]; !ok {
-			err = ErrTsNotFound
+		entry = jsonEntry{}
+		entry.Value = valuePtr
+		err := json.Unmarshal(bytes, &entry)
+		if err != nil {
+			return 0, err
 		}
-		return
+		j.cachedData[key] = entry
 	}
 
-	err = ErrNotFound
-	return
+	// Using reflection to indirectly set the value passed as reference
+	reflect.Indirect(rv).Set(reflect.Indirect(reflect.ValueOf(entry.Value)))
+
+	return entry.Timestamp, nil
 }
 
-// Delete removes the key entry
-func (c *inMemoryStore) Delete(name string) {
-	delete(c.Data, name)
-	delete(c.Timestamps, name)
+// flushcache marshalls all the cached data into JSON, ready to be stored into disk
+func (j *inMemoryStore) flushCache() error {
+	for k, v := range j.cachedData {
+		bytes, err := json.Marshal(v)
+		if err != nil {
+			return err
+		}
+		j.Data[k] = bytes
+	}
+	j.cachedData = make(map[string]jsonEntry)
+	return nil
 }
 
-// Set adds a value into the store and it also stores the current timestamp.
-func (c *inMemoryStore) Set(name string, value interface{}) int64 {
-	c.Data[name] = value
-	c.Timestamps[name] = now().Unix()
-	return c.Timestamps[name]
+func (j *fileStore) loadFromDisk() {
+	bytes, err := ioutil.ReadFile(j.path)
+	if err != nil {
+		j.ilog.Debugf("can't read %q: %s. Ignoring", j.path, err.Error())
+	} else {
+		err = json.Unmarshal(bytes, j)
+		if err != nil {
+			j.ilog.Debugf("can't unmarshall %q: %s. Ignoring", j.path, err.Error())
```

Comment:
return error instead of log please

Reply:
Same comment as: https://github.com/newrelic/infra-integrations-sdk/pull/92#discussion_r190528470